### ISSUE TITLE
Browsing history

### ISSUE THREAD

**Issue by JunnanZ** [moosotc/llpp#64]
A common scenario: After I go to a bookmark, I can go back to my previous position by `backspace` or `alt+leftarrow`. However, if I want to go to that bookmark again, I can't go there by `alt+rightarrow`. In fact, that bookmark is not even in the history. Is this a bug or just how it works?

**Comment by moosotc** [comment]
Junnan <notifications@github.com> writes:

> A common scenario: After I go to a bookmark, I can go back to my
> previous position by `backspace` or `alt+leftarrow`. However, if I
> want to go to that bookmark again, I can't go there by
> `alt+rightarrow`. In fact, that bookmark is not even in the
> history. Is this a bug or just how it works?

Can't say I understand what is that you are asking. Perhaps you can
provide an exact reproduction scenario, i.e. the document, what keys are
pressed, what's the expected and obtained result.

-- 
mailto:moosotc@gmail.com

**Comment by JunnanZ** [comment]
@moosotc Sorry I was not being clear. 

To reproduce, open any pdf file longer than 1 page, press `G` to go to the last page, and then press `alt+leftarrow` to go back to the first page. Now what I want is to press `alt+rightarrow` to go to the last page again, since I browsed the last page so that should be in my browsing "history". However, neither `alt+leftarrow` or `alt+rightarrow` takes me anywhere. It's like the last page was not in the "history".

**Comment by moosotc** [comment]
Junnan <notifications@github.com> writes:

> @moosotc Sorry I was not being clear. 
>
> To reproduce, open any pdf file longer than 1 page, press `G` to go to
> the last page, and then press `alt+leftarrow` to go back to the first
> page. Now what I want is to press `alt+rightarrow` to go to the last
> page again, since I browsed the last page so that should be in my
> browsing "history". However, neither `alt+leftarrow` or
> `alt+rightarrow` takes me anywhere. It's like the last page was not in
> the "history".

Thanks, this is by mis-design I'm afraid, going back/forwards does not
mark current location in history. Whole "history" navigation is rather
fragile and hackish, also I don't have a clear idea how to "fix" that.

-- 
mailto:moosotc@gmail.com

**Comment by JunnanZ** [comment]
Yes, that confuses me a lot. By what criteria does llpp mark a page in history? It seems pretty random to me. Maybe you could mark a page in history if the page is arrived by a "jump", i.e., go to a bookmark, go to an outline entry, go to specific page, etc.?

**Comment by moosotc** [comment]
Junnan <notifications@github.com> writes:

> Yes, that confuses me a lot. By what criteria does llpp mark a page in
> history? It seems pretty random to me. Maybe you could mark a page in
> history if the page is arrived by a "jump", i.e., go to a bookmark, go
> to an outline entry, go to specific page, etc.?

That's precisely what it does, it's just that going back/forward is not
considered a mark-in-history event.

-- 
mailto:moosotc@gmail.com

**Comment by JunnanZ** [comment]
I get it.

**Comment by JunnanZ** [comment]
Just found a workaround. Whenever I go to a bookmark and then want to go back in history, I press `g` first to register this page in history.

**Comment by moosotc** [comment]
Junnan <notifications@github.com> writes:

> Just found a workaround. Whenever I go to a bookmark and then want to
> go back in history, I press `g` first to register this page in
> history.

You keep mentioning bookmarks... "'" bookmarks or are you mixing up
concepts?

-- 
mailto:moosotc@gmail.com

**Comment by JunnanZ** [comment]
No. For example, I'm reading a paper and want to take a look at the references, which I saved as a bookmark. So I go to that bookmark and then go back in history to where I was reading previously. I may want to check the references again, so it would be convenient if I can go forward in history to the references.

**Comment by moosotc** [comment]
Junnan <notifications@github.com> writes:

> No. For example, I'm reading a paper and want to take a look at the
> references, which I saved as a bookmark. So I go to that bookmark and
> then go back in history to where I was reading previously. I may want
> to check the references again, so it would be convenient if I can go
> forward in history to the references.

You should be able to, the fact that you can't signifies that it's not
trivial :)

-- 
mailto:moosotc@gmail.com

**Comment by JunnanZ** [comment]
I applied the patch and rebuilt, but it seems that there is no difference. What's the expected behavior?

**Comment by moosotc** [comment]
Junnan <notifications@github.com> writes:

> I applied the patch and rebuilt, but it seems that there is no
> difference. What's the expected behavior?

Sorry, brain fart, back to the drawing board.

-- 
mailto:moosotc@gmail.com

**Comment by moosotc** [comment]
Junnan <notifications@github.com> writes:

> I applied the patch and rebuilt, but it seems that there is no
> difference. What's the expected behavior?

Can you try following?

diff --git a/main.ml b/main.ml
index a752ca4..c1fb770 100644
--- a/main.ml
+++ b/main.ml
@@ -4766,6 +4766,9 @@ let viewkeyboard key mask =
      end;
 
   | Backspace ->
+     let rc = state.hists.nav.rc in
+     addnav ();
+     state.hists.nav.rc <- rc;
      gotoghyll (getnav ~-1)
 
   | Ascii 'o' ->
@@ -5160,8 +5163,14 @@ let viewkeyboard key mask =
      gotoghyll (clamp state.maxy)
 
   | Right when Wsi.withalt mask ->
+     let rc = state.hists.nav.rc in
+     addnav ();
+     state.hists.nav.rc <- rc;
      gotoghyll (getnav 1)
   | Left when Wsi.withalt mask ->
+     let rc = state.hists.nav.rc in
+     addnav ();
+     state.hists.nav.rc <- rc;
      gotoghyll (getnav ~-1)
 
   | Ascii 'r' ->

-- 
mailto:moosotc@gmail.com

**Comment by JunnanZ** [comment]
It works! Thank you so much.

**Comment by JunnanZ** [comment]
I notice some subtleties of the new behaviors. To reproduce: open a pdf file longer than 1 page, press `G`, then press `uparrow` to move a little bit, and then press `alt+leftarrow` to go back. Now, if I press `alt+rightarrow`, it will go to the position after I moved up instead of the position after I pressed `G`. This is a desired behavior I think, because it remembers where I was reading instead of where the last jump led me to.

However, this seems to be the only case history navigation remembers the position. In other cases, for example, after I go back from the last page, I press `alt+rightarrow` to go to the last page again, now if I move up it will not remember my position after I `alt+left` and `alt+right`. It seems that `llpp` marks a position in history with `alt+left` or `alt+right` only if I go forward/backward right after a jump (go to page, go to a bookmark, go to last page, etc.). 

Bottom line is, I think it would be more consistent if `llpp` treats every going backward/forward a "mark-in-history" event.

**Comment by moosotc** [comment]
Junnan <notifications@github.com> writes:

> I notice some subtleties of the new behaviors. To reproduce: open a
> pdf file longer than 1 page, press `G`, then press `uparrow` to move a
> little bit, and then press `alt+leftarrow` to go back. Now, if I press
> `alt+rightarrow`, it will go to the position after I moved up instead
> of the position after I pressed `G`. This is a desired behavior I
> think, because it remembers where I was reading instead of where the
> last jump led me to.
>
> However, this seems to be the only case history navigation remembers
> the position. In other cases, for example, after I go back from the
> last page, I press `alt+rightarrow` to go to the last page again, now
> if I move up it will not remember my position after I `alt+left` and
> `alt+right`. It seems that `llpp` marks a position in history with
> `alt+left` or `alt+right` only if I go forward/backward right after a
> jump (go to page, go to a bookmark, go to last page, etc.).
>
> Bottom line is, I think it would be more consistent if `llpp` treats
> every going backward/forward a "mark-in-history" event.

Hmm.. it does do that. Every alt-left/right-arrow|backspace puts the
position at the point of key press into history. This tells me that
proper description/model of the navigational behavior is needed and a
way to check that the code follows the model.

It's clear that my approach of writing things on a mental napkin and
hoping it will do the right thing, whatever that might be, fails me once
again.

-- 
mailto:moosotc@gmail.com

**Comment by moosotc** [comment]
Junnan <notifications@github.com> writes:

> I notice some subtleties of the new behaviors. To reproduce: open a
> pdf file longer than 1 page, press `G`, then press `uparrow` to move a
> little bit, and then press `alt+leftarrow` to go back. Now, if I press
> `alt+rightarrow`, it will go to the position after I moved up instead
> of the position after I pressed `G`. This is a desired behavior I
> think, because it remembers where I was reading instead of where the
> last jump led me to.
>
> However, this seems to be the only case history navigation remembers
> the position. In other cases, for example, after I go back from the
> last page, I press `alt+rightarrow` to go to the last page again, now
> if I move up it will not remember my position after I `alt+left` and
> `alt+right`. It seems that `llpp` marks a position in history with
> `alt+left` or `alt+right` only if I go forward/backward right after a
> jump (go to page, go to a bookmark, go to last page, etc.).
>
> Bottom line is, I think it would be more consistent if `llpp` treats
> every going backward/forward a "mark-in-history" event.

Following let's one visualise the navigational history (abusing history
'H') and it's a bloody mess :(

diff --git a/main.ml b/main.ml
index c84288b..881f391 100644
--- a/main.ml
+++ b/main.ml
@@ -3267,7 +3267,7 @@ class outlinelistview ~zebra ~source =
          super#key key mask
   end;;
 
-let genhistoutlines () =
+let _genhistoutlines () =
   Config.gethist ()
   |> List.sort (fun (_, c1, _, _, _, _) (_, c2, _, _, _, _) ->
          compare c2.lastvisit c1.lastvisit)
@@ -3278,6 +3278,27 @@ let genhistoutlines () =
        )
 ;;
 
+let genhistoutlines () =
+  let l = Array.to_list state.hists.nav.store in
+  let b = state.hists.nav in
+  (Printf.sprintf "cap=%d len=%d rc=%d wc=%d"
+                  (Array.length b.store) b.len b.rc b.wc, 0, Onone)
+  ::
+    List.mapi (fun i c ->
+         let pageno, top, _ = c in
+         let s =
+           let s = if i = b.rc then ">" else "" in
+           let s = if i = b.wc then s ^ "<" else s in
+           s
+         in
+         (
+           (Printf.sprintf "%4d: page %4d top %2.1f %s" i pageno top s),
+           1,
+           Oanchor c
+         )
+    ) l
+;;
+
 let gotohist (path, c, bookmarks, x, anchor, origin) =
   Config.save leavebirdseye;
   state.anchor <- anchor;

-- 
mailto:moosotc@gmail.com

**Comment by JunnanZ** [comment]
I don't quite understand how to utilize this...

**Comment by moosotc** [comment]
Junnan <notifications@github.com> writes:

> I don't quite understand how to utilize this...

Pressing 'H' will show the navigational instead of visited history.

Here's a slightly better version, things are still very confusing
though.

diff --git a/main.ml b/main.ml
index c84288b..f9fbf16 100644
--- a/main.ml
+++ b/main.ml
@@ -3268,14 +3268,35 @@ class outlinelistview ~zebra ~source =
   end;;
 
 let genhistoutlines () =
-  Config.gethist ()
-  |> List.sort (fun (_, c1, _, _, _, _) (_, c2, _, _, _, _) ->
-         compare c2.lastvisit c1.lastvisit)
-  |> List.map (fun ((path, c, _, _, _, origin) as hist) ->
-         let path = if nonemptystr origin then origin else path in
-         let base = mbtoutf8 @@ Filename.basename path in
-         (base ^ "\000" ^ c.title, 1, Ohistory hist)
-       )
+  if false
+  then
+    Config.gethist ()
+    |> List.sort (fun (_, c1, _, _, _, _) (_, c2, _, _, _, _) ->
+           compare c2.lastvisit c1.lastvisit)
+    |> List.map (fun ((path, c, _, _, _, origin) as hist) ->
+           let path = if nonemptystr origin then origin else path in
+           let base = mbtoutf8 @@ Filename.basename path in
+           (base ^ "\000" ^ c.title, 1, Ohistory hist)
+         )
+  else
+    let l = Array.to_list state.hists.nav.store in
+    let b = state.hists.nav in
+    (Printf.sprintf "cap=%d len=%d rc=%d wc=%d"
+                    (Array.length b.store) b.len b.rc b.wc, 0, Onone)
+    ::
+      List.mapi (fun i c ->
+          let pageno, top, _ = c in
+          let s =
+            let s = if i = b.rc then ">" else "" in
+            let s = if i = b.wc then s ^ "<" else s in
+            s
+          in
+          (
+            (Printf.sprintf "%4d: page %4d top %2.1f %s" i pageno top s),
+            1,
+            Oanchor c
+          )
+        ) l
 ;;
 
 let gotohist (path, c, bookmarks, x, anchor, origin) =

-- 
mailto:moosotc@gmail.com

**Comment by JunnanZ** [comment]
Yes, it is confusing. Is there any way to make history navigation a mark-history event?

**Comment by moosotc** [comment]
Junnan <notifications@github.com> writes:

> Yes, it is confusing. Is there any way to make history navigation a
> mark-history event?

Not sure what you mean by that.
-- 
mailto:moosotc@gmail.com

**Comment by JunnanZ** [comment]
I mean, like you said, "every alt-left/right-arrow|backspace puts the position at the point of key press into history".

**Comment by moosotc** [comment]
Junnan <notifications@github.com> writes:

> I mean, like you said, "every alt-left/right-arrow|backspace puts the
> position at the point of key press into history".

That's what it does now... I'm confused.

-- 
mailto:moosotc@gmail.com

**Comment by JunnanZ** [comment]
Oh I get it! Thank you!
